ethers-core/src/utils/mod.rs

```diff
@@ -515,14 +515,20 @@ where
 {
     #[derive(Deserialize)]
     #[serde(untagged)]
-    enum IntOrHex {
+    enum IntOrString {
         Int(serde_json::Number),
-        Hex(String),
+        JsonString(String),
     }
 
-    match IntOrHex::deserialize(deserializer)? {
-        IntOrHex::Hex(s) => U256::from_str(s.as_str()).map_err(serde::de::Error::custom),
-        IntOrHex::Int(n) => U256::from_dec_str(&n.to_string()).map_err(serde::de::Error::custom),
+    match IntOrString::deserialize(deserializer)? {
+        IntOrString::JsonString(s) => {
+            if s.starts_with("0x") {
+                U256::from_str(s.as_str()).map_err(serde::de::Error::custom)
+            } else {
+                U256::from_dec_str(&s).map_err(serde::de::Error::custom)
+            }
```

Comment:
cc @prestwich for ethers-next, you wouldn't believe how much time pinpointing this took us